controller/machine/utils.go

```diff
@@ -252,3 +253,52 @@ func waitUntilSSHKey(machineDir string, machine *v3.Machine) error {
 		return nil
 	}
 }
+
+func checkSingleNode(machine *v3.Machine, clusters v3.ClusterInterface, machineInterface v3.MachineInterface) (bool, error) {
+	var (
+		etcd, controlplane bool
+	)
+	clusterName := machine.Spec.ClusterName
+	if len(clusterName) == 0 {
+		return false, nil
+	}
+	namespace := machine.Namespace
+	if len(namespace) == 0 {
+		return false, nil
+	}
+	clusterObj, err := clusters.Get(clusterName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+	if clusterObj.Spec.RancherKubernetesEngineConfig == nil {
+		return false, nil
+	}
+	if len(machine.Spec.Role) == 1 && machine.Spec.Role[0] == "worker" {
+		return false, nil
+	}
+	machines, err := machineInterface.Controller().Lister().List(namespace, labels.Everything())
```

Comment:
don't think we should check the machines. When the machine gets assigned to the cluster, the cluster spec gets updated. So may be it would make sense to check the cluster directly